src/query/pipeline/core/src/finished_chain.rs

```diff
@@ -0,0 +1,388 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+use std::fmt::Write;
+use std::panic::Location;
+use std::time::Duration;
+use std::time::Instant;
+
+use databend_common_base::runtime::catch_unwind;
+use databend_common_exception::Result;
+use log::info;
+
+use crate::PlanProfile;
+
+pub struct ExecutionInfo {
+    pub res: Result<()>,
+    pub profiling: Vec<PlanProfile>,
+}
+
+impl ExecutionInfo {
+    pub fn create(res: Result<()>, profiling: Vec<PlanProfile>) -> ExecutionInfo {
+        ExecutionInfo { res, profiling }
+    }
+}
+
+pub trait Callback: Send + Sync + 'static {
+    fn always_call(&self) -> bool {
+        false
+    }
+
+    fn apply(self: Box<Self>, info: &ExecutionInfo) -> Result<()>;
+}
+
+struct ApplyState {
+    is_always: bool,
+    successfully: bool,
+    elapsed: Duration,
+    location: &'static Location<'static>,
+}
+
+pub struct FinishedCallbackChain {
+    chain: VecDeque<(&'static Location<'static>, Box<dyn Callback>)>,
+}
+
+impl FinishedCallbackChain {
+    pub fn create() -> FinishedCallbackChain {
+        FinishedCallbackChain {
+            chain: VecDeque::new(),
+        }
+    }
+
+    pub fn push_front(&mut self, location: &'static Location<'static>, f: Box<dyn Callback>) {
+        self.chain.push_front((location, f))
+    }
+
+    pub fn push_back(&mut self, location: &'static Location<'static>, f: Box<dyn Callback>) {
+        self.chain.push_back((location, f))
+    }
+
+    pub fn apply(&mut self, mut info: ExecutionInfo) -> Result<()> {
+        let chain = std::mem::take(&mut self.chain);
+
+        let mut states = Vec::with_capacity(chain.len());
+        let mut callbacks = Vec::with_capacity(chain.len());
+        let mut always_callbacks = Vec::with_capacity(chain.len());
+
+        for (location, callback) in chain.into_iter() {
+            if !callback.always_call() {
+                callbacks.push((location, callback));
+            } else {
+                always_callbacks.push((location, callback));
+            }
+        }
+
+        for (location, callback) in callbacks {
+            let instant = Instant::now();
+            if let Err(cause) = callback.apply(&info) {
+                states.push(ApplyState {
+                    location,
+                    is_always: false,
+                    successfully: false,
+                    elapsed: instant.elapsed(),
+                });
+
+                info.res = Err(cause.clone());
+                Self::apply_always(info, states, always_callbacks);
+
+                return Err(cause);
```

Comment:
logging the locations of callbacks that have not run may help debug too.
otherwise, some developers may be unaware of its existence, making it difficult to associate the program’s erroneous behavior with it.